Datadog container panic after adding prometheus scrape

<!--
If you suspect your issue is a bug, please attach the output of the Agent status page,
see https://docs.datadoghq.com/agent/faq/agent-commands/#agent-information
-->

**Output of the info page (if this is a bug)**

This is from a running container since I can't get this from the one which panics:

```
Getting the status from the agent.

===============
Agent (v6.14.1)
===============

  Status date: 2019-10-01 08:54:31.050245 UTC
  Agent start: 2019-10-01 07:10:34.496840 UTC
  Pid: 335
  Go Version: go1.12.9
  Python Version: 2.7.16
  Check Runners: 4
  Log Level: info

  Paths
  =====
    Config File: /etc/datadog-agent/datadog.yaml
    conf.d: /etc/datadog-agent/conf.d
    checks.d: /etc/datadog-agent/checks.d

  Clocks
  ======
    NTP offset: 934µs
    System UTC time: 2019-10-01 08:54:31.050245 UTC

  Host Info
  =========
    bootTime: 2019-07-12 21:49:56.000000 UTC
    kernelVersion: 4.14.114-103.97.amzn2.x86_64
    os: linux
    platform: debian
    platformFamily: debian
    platformVersion: 10.1
    procs: 68
    uptime: 1929h20m44s

  Hostnames
  =========
    ec2-hostname: xxxxx
    hostname: i-xxxx
    instance-id: i-xxxx
    socket-fqdn: 49275f2a3046
    socket-hostname: 49275f2a3046
    hostname provider: aws
    unused hostname providers:
      configuration/environment: hostname is empty
      gce: unable to retrieve hostname from GCE: status code 404 trying to GET http://169.254.169.254/computeMetadata/v1/instance/hostname

=========
Collector
=========



  Running Checks
  ==============
    
    cpu
    ---
      Instance ID: cpu [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/cpu.d/conf.yaml.default
      Total Runs: 415
      Metric Samples: Last Run: 6, Total: 2,484
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
    
    disk (2.5.0)
    ------------
      Instance ID: disk:e5dffb8bef24336f [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/disk.d/conf.yaml.default
      Total Runs: 415
      Metric Samples: Last Run: 168, Total: 69,720
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 103ms
      
    
    docker
    ------
      Instance ID: docker [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/docker.d/conf.yaml.default
      Total Runs: 415
      Metric Samples: Last Run: 98, Total: 56,040
      Events: Last Run: 0, Total: 876
      Service Checks: Last Run: 1, Total: 415
      Average Execution Time : 324ms
      
    
    file_handle
    -----------
      Instance ID: file_handle [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/file_handle.d/conf.yaml.default
      Total Runs: 415
      Metric Samples: Last Run: 5, Total: 2,075
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
    
    io
    --
      Instance ID: io [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/io.d/conf.yaml.default
      Total Runs: 415
      Metric Samples: Last Run: 52, Total: 21,544
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
    
    load
    ----
      Instance ID: load [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/load.d/conf.yaml.default
      Total Runs: 415
      Metric Samples: Last Run: 6, Total: 2,490
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
    
    memory
    ------
      Instance ID: memory [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/memory.d/conf.yaml.default
      Total Runs: 416
      Metric Samples: Last Run: 17, Total: 7,072
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
    
    network (1.11.4)
    ----------------
      Instance ID: network:e0204ad63d43c949 [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/network.d/conf.yaml.default
      Total Runs: 415
      Metric Samples: Last Run: 67, Total: 35,611
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 3ms
      
    
    ntp
    ---
      Instance ID: ntp:d884b5186b651429 [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/ntp.d/conf.yaml.default
      Total Runs: 416
      Metric Samples: Last Run: 1, Total: 416
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 1, Total: 416
      Average Execution Time : 0s
      
    
    prometheus (3.2.0)
    ------------------
      Instance ID: prometheus:pm:e6f0a20c606a9558 [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/prometheus.d/conf.yaml
      Total Runs: 416
      Metric Samples: Last Run: 77, Total: 20,991
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 1, Total: 416
      Average Execution Time : 14ms
      
    
    uptime
    ------
      Instance ID: uptime [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/uptime.d/conf.yaml.default
      Total Runs: 415
      Metric Samples: Last Run: 1, Total: 415
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
========
JMXFetch
========

  Initialized checks
  ==================
    no checks
    
  Failed checks
  =============
    no checks
    
=========
Forwarder
=========

  Transactions
  ============
    CheckRunsV1: 415
    Dropped: 0
    DroppedOnInput: 0
    Events: 0
    HostMetadata: 0
    IntakeV1: 328
    Metadata: 0
    Requeued: 0
    Retried: 0
    RetryQueueSize: 0
    Series: 0
    ServiceChecks: 0
    SketchSeries: 0
    Success: 1,158
    TimeseriesV1: 415

  API Keys status
  ===============
    API key ending with 600f9: API Key valid

==========
Endpoints
==========
  https://app.datadoghq.com - API Key ending with:
      - 600f9

==========
Logs Agent
==========

  Logs Agent is not running

=========
Aggregator
=========
  Checks Metric Sample: 227,175
  Dogstatsd Metric Sample: 1
  Event: 877
  Events Flushed: 877
  Number Of Flushes: 415
  Series Flushed: 157,407
  Service Check: 5,406
  Service Checks Flushed: 5,812

=========
DogStatsD
=========
  Event Packets: 0
  Event Parse Errors: 0
  Metric Packets: 0
  Metric Parse Errors: 0
  Service Check Packets: 0
  Service Check Parse Errors: 0
  Udp Bytes: 0
  Udp Packet Reading Errors: 0
  Udp Packets: 0
  Uds Bytes: 0
  Uds Origin Detection Errors: 0
  Uds Packet Reading Errors: 0
  Uds Packets: 0
```

**Describe what happened:**
The dd-agent container fails to start sometimes and fails with a go panic. It does not happen every time. Here is the log output:
```
[...]
2019-10-01 07:10:14 UTC | CORE | INFO | (pkg/dogstatsd/listeners/udp.go:79 in Listen) | dogstatsd-udp: starting to listen on [::]:8125
2019-10-01 07:10:14 UTC | CORE | INFO | (cmd/agent/app/start.go:219 in StartAgent) | logs-agent disabled
2019-10-01 07:10:14 UTC | CORE | INFO | (pkg/tagger/tagger.go:152 in tryCollectors) | ecs tag collector successfully started
2019-10-01 07:10:14 UTC | CORE | INFO | (pkg/tagger/tagger.go:152 in tryCollectors) | docker tag collector successfully started
2019-10-01 07:10:14 UTC | CORE | INFO | (pkg/collector/runner/runner.go:92 in NewRunner) | Runner started with 4 workers.
2019-10-01 07:10:14 UTC | CORE | INFO | (pkg/collector/python/init.go:274 in Initialize) | Initializing rtloader with python2 /opt/datadog-agent/embedded
2019-10-01 07:10:15 UTC | CORE | INFO | (pkg/collector/python/datadog_agent.go:118 in LogMessage) | - | (ddyaml.py:123) | monkey patching yaml.load...
2019-10-01 07:10:15 UTC | CORE | INFO | (pkg/collector/python/datadog_agent.go:118 in LogMessage) | - | (ddyaml.py:127) | monkey patching yaml.load_all...
2019-10-01 07:10:15 UTC | CORE | INFO | (pkg/collector/python/datadog_agent.go:118 in LogMessage) | - | (ddyaml.py:131) | monkey patching yaml.dump_all... (affects all yaml dump operations)
2019-10-01 07:10:16 UTC | CORE | INFO | (pkg/collector/collector.go:57 in NewCollector) | Embedding Python 2.7.16 (default, Sep 26 2019, 12:11:18) [GCC 4.7.2]
2019-10-01 07:10:16 UTC | CORE | INFO | (cmd/agent/common/autoconfig.go:71 in SetupAutoConfig) | Registering docker config provider polled every 1s
2019-10-01 07:10:16 UTC | CORE | INFO | (pkg/autodiscovery/autoconfig.go:354 in initListenerCandidates) | docker listener successfully started
2019-10-01 07:10:16 UTC | CORE | INFO | (pkg/autodiscovery/providers/file.go:72 in Collect) | file: searching for configuration files at: /etc/datadog-agent/conf.d
2019-10-01 07:10:16 UTC | CORE | INFO | (pkg/autodiscovery/providers/file.go:72 in Collect) | file: searching for configuration files at: /opt/datadog-agent/bin/agent/dist/conf.d
2019-10-01 07:10:16 UTC | CORE | WARN | (pkg/autodiscovery/providers/file.go:76 in Collect) | Skipping, open /opt/datadog-agent/bin/agent/dist/conf.d: no such file or directory
2019-10-01 07:10:16 UTC | CORE | INFO | (pkg/autodiscovery/providers/file.go:72 in Collect) | file: searching for configuration files at:
2019-10-01 07:10:16 UTC | CORE | WARN | (pkg/autodiscovery/providers/file.go:76 in Collect) | Skipping, open : no such file or directory
panic: assignment to entry in nil map

goroutine 152 [running]:
github.com/DataDog/datadog-agent/pkg/autodiscovery/providers.(*DockerConfigProvider).listen(0xc0007d48c0)
        /.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/providers/docker.go:114 +0x3d0
created by github.com/DataDog/datadog-agent/pkg/autodiscovery/providers.(*DockerConfigProvider).Collect
        /.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/providers/docker.go:60 +0x1b5
AGENT EXITED WITH CODE 2, SIGNAL 0, KILLING CONTAINER
2019-10-01 07:10:16 UTC | PROCESS | CRITICAL | (pkg/process/util/signal_nowindows.go:21 in HandleSignals) | Caught signal 'terminated'; terminating.
2019-10-01 07:10:16 UTC | TRACE | INFO | (main.go:23 in handleSignal) | received signal 15 (terminated)
system-probe exited with code 256, signal 15, restarting in 2 seconds
process-agent exited with code 0, disabling
trace-agent exited with code 0, disabling
```

**Describe what you expected:**
dd-agent container should not fail

**Steps to reproduce the issue:**
It seems to have started to happen after we added the Prometheus endpoint scraping configuration. This was added to the file `/etc/datadog-agent/conf.d/prometheus.d/conf.yaml`, and configuration is like so:

```
init_config:

instances:

    ## @param prometheus_url - string - required
    ## The URL where your application metrics are exposed by Prometheus.
    #
  - prometheus_url: http://_my_host_name_:9102/metrics

    ## @param namespace - string - required
    ## The namespace to be appended before all metrics namespace
    #
    namespace: 'pm'

    ## @param metrics - list of key:value elements - required
    ## List of `<METRIC_TO_FETCH>: <NEW_METRIC_NAME>` for metrics to be fetched from the prometheus endpoint.
    ## <NEW_METRIC_NAME> is optional. It transforms the name in Datadog if set.
    ## This list should contain at least one metric
    #
    metrics:
      - mymetric1
      - mymetric2
[...]
```

**Additional environment details (Operating System, Cloud provider, etc):**
The problem is seen on AWS ECS and docker service running on an AWS instance.

--- a/pkg/autodiscovery/providers/docker.go
+++ b/pkg/autodiscovery/providers/docker.go
@@ -57,7 +57,6 @@ func (d *DockerConfigProvider) Collect() ([]integration.Config, error) {
 		if err != nil {
 			return []integration.Config{}, err
 		}
-		go d.listen()
 	}
 
 	var containers map[string]map[string]string
@@ -78,6 +77,9 @@ func (d *DockerConfigProvider) Collect() ([]integration.Config, error) {
 	d.syncCounter += 1
 	d.upToDate = true
 	d.Unlock()
+
+	// start listening after the first collection to avoid race in cache map init
+	go d.listen()
 
 	return parseDockerLabels(containers)
 }

--- a/releasenotes/notes/fix-docker-crash-d18d71e7b79c34bd.yaml
+++ b/releasenotes/notes/fix-docker-crash-d18d71e7b79c34bd.yaml
@@ -1,0 +1,4 @@
+---
+fixes:
+  - |
+    Fix a bug where the agent would crash when using the docker autodiscovery config provider.

